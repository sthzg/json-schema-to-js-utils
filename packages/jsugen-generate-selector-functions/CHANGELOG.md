# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.8.0](https://github.com/sthzg/jsugen/compare/v0.3.0...v0.8.0) (2019-05-07)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





# [0.7.0](https://github.com/sthzg/jsugen/compare/v0.3.0...v0.7.0) (2019-05-07)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





# [0.6.0](https://github.com/sthzg/jsugen/compare/v0.3.0...v0.6.0) (2019-05-07)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





# [0.5.0](https://github.com/sthzg/jsugen/compare/v0.3.0...v0.5.0) (2019-05-07)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





# [0.3.0](https://github.com/sthzg/jsugen/compare/v0.2.0...v0.3.0) (2019-05-06)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





# [0.2.0](https://github.com/sthzg/jsugen/compare/v0.1.5...v0.2.0) (2019-05-05)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





## [0.1.5](https://github.com/sthzg/jsugen/compare/v0.1.4...v0.1.5) (2019-05-02)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





## [0.1.4](https://github.com/sthzg/jsugen/compare/v0.1.3...v0.1.4) (2019-05-02)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





## [0.1.3](https://github.com/sthzg/jsugen/compare/v0.1.2...v0.1.3) (2019-05-01)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





## [0.1.2](https://github.com/sthzg/jsugen/compare/v0.1.1...v0.1.2) (2019-04-27)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





## [0.1.1](https://github.com/sthzg/jsugen/compare/v0.1.0...v0.1.1) (2019-02-28)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





# [0.1.0](https://github.com/sthzg/jsugen/compare/v0.0.4...v0.1.0) (2019-02-28)


### Features

* support passing default return values to selectors ([bff3629](https://github.com/sthzg/jsugen/commit/bff3629))





## [0.0.4](https://github.com/sthzg/jsugen/compare/v0.0.3...v0.0.4) (2019-02-28)

**Note:** Version bump only for package @sthzg/jsugen-generate-selector-functions





## [0.0.3](https://github.com/sthzg/jsugen/compare/v0.0.3-beta.1...v0.0.3) (2019-02-28)


### Bug Fixes

* make build chain work and publish ([fc00c4e](https://github.com/sthzg/jsugen/commit/fc00c4e))
