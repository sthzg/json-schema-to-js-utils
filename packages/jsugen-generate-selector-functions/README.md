# `jsugen-generate-selector-functions`

## Usage

```
const jsugenGenerateSelectorFunctions = require('jsugen-generate-selector-functions');

// TODO: DEMONSTRATE API
```
