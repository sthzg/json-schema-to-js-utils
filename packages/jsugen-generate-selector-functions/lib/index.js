import { createWriteStream } from 'fs';
import { pipeline as pipelineFn } from 'stream';
import { promisify } from 'util';
import {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  FilterSchemaKeywordsTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  PrependToFileTransform,
  PrettierTransform,
  StdoutWritable,
} from '@sthzg/jsugen-core/lib/streams';
import {
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_GET_IMPORT,
  DEFAULT_PRETTIER_OPTIONS,
} from '@sthzg/jsugen-core/lib/constants';
import selectorFunctionModule from './selector.tpl';

const pipeline = promisify(pipelineFn);

/**
 * Generates selector functions for all values inside the JSON schema.
 *
 * Selectors are functions that pick a value from an object in a failsafe way or return undefined if
 * the property is not defined on the input object, e.g.
 *
 *     function byPersonAddressStreet(person) {
 *       return get(person, 'address.street');
 *     }
 */
function generateSelectorsModule({ schema, out }) {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  return pipeline(
    jsonSchemaReadable(schema),
    new FilterSchemaKeywordsTransform(),
    new BuildObjectPathsTransform(),
    new CompileToTemplateTransform(selectorFunctionModule),
    new MemorySinkTransform(),
    new PrependToFileTransform([DEFAULT_FILE_DOCSTRING, DEFAULT_GET_IMPORT]),
    new PrettierTransform(DEFAULT_PRETTIER_OPTIONS),
    write
  );
}

export default generateSelectorsModule;
