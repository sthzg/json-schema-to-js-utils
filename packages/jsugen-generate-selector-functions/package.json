{
  "name": "@sthzg/jsugen-generate-selector-functions",
  "version": "0.0.2",
  "description": "> TODO: description",
  "author": "Stephan Herzog <sthzgvie@gmail.com>",
  "homepage": "https://github.com/sthzg/jsugen#readme",
  "license": "MIT",
  "main": "lib/index.js",
  "directories": {
    "lib": "lib"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sthzg/jsugen.git"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "bugs": {
    "url": "https://github.com/sthzg/jsugen/issues"
  },
  "dependencies": {
    "@sthzg/jsugen-core": "^0.0.2",
    "lodash.camelcase": "^4.3.0"
  }
}
