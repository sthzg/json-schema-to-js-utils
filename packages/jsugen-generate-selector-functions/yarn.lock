# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


lodash.camelcase@^4.3.0:
  version "4.3.0"
  resolved "https://registry.yarnpkg.com/lodash.camelcase/-/lodash.camelcase-4.3.0.tgz#b28aa6288a2b9fc651035c7711f65ab6190331a6"
  integrity sha1-soqmKIorn8ZRA1x3EfZathkDMaY=
