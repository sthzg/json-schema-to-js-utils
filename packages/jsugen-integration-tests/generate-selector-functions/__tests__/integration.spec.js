import { generate } from '@sthzg/jsugen-generate-selector-functions';
import {
  CONTACT_SCHEMA_PATH,
  EXAMPLE_SCHEMA_PATH,
  DryRunWriteConfig as writeConfig,
} from '../../testUtils/helpers';

describe('generate-selector-functions', () => {
  test('should pass full integration test', done => {
    expect.assertions(1);

    generate({ sourceFile: EXAMPLE_SCHEMA_PATH, writeConfig }).subscribe(
      output => {
        expect(output).toMatchSnapshot();
        done();
      },
    );
  });

  test('should pass example.contact.json integration test', done => {
    expect.assertions(1);

    generate({ sourceFile: CONTACT_SCHEMA_PATH, writeConfig }).subscribe(
      output => {
        expect(output).toMatchSnapshot();
        done();
      },
    );
  });
});
