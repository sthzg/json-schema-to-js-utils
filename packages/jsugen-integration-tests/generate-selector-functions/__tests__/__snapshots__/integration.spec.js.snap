// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-selector-functions should pass example.contact.json integration test 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import get from 'lodash/get';

export function byId(property, defaultReturn) {
  return get(property, ['id'], defaultReturn);
}

export function byFirstName(property, defaultReturn) {
  return get(property, ['firstName'], defaultReturn);
}

export function byLastName(property, defaultReturn) {
  return get(property, ['lastName'], defaultReturn);
}

export function byGender(property, defaultReturn) {
  return get(property, ['gender'], defaultReturn);
}

export function byPhoneNthType(property, phoneIndex = 0, defaultReturn) {
  return get(property, ['phone', phoneIndex, 'type'], defaultReturn);
}

export function byPhoneNthNumber(property, phoneIndex = 0, defaultReturn) {
  return get(property, ['phone', phoneIndex, 'number'], defaultReturn);
}

export function byPhone(property, defaultReturn) {
  return get(property, ['phone'], defaultReturn);
}

export function byPhoneNth(property, phoneIndex = 0, defaultReturn) {
  return get(property, ['phone', phoneIndex], defaultReturn);
}
"
`;

exports[`generate-selector-functions should pass full integration test 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import get from 'lodash/get';

export function byLevel1Member1(property, defaultReturn) {
  return get(property, ['level1Member1'], defaultReturn);
}

export function byLevel1Member2(property, defaultReturn) {
  return get(property, ['level1Member2'], defaultReturn);
}

export function byLevel1Member3(property, defaultReturn) {
  return get(property, ['level1Member3'], defaultReturn);
}

export function byLevel1Member4(property, defaultReturn) {
  return get(property, ['level1Member4'], defaultReturn);
}

export function byLevel1Member5(property, defaultReturn) {
  return get(property, ['level1Member5'], defaultReturn);
}

export function byLevel1Member5Nth(
  property,
  level1Member5Index = 0,
  defaultReturn,
) {
  return get(property, ['level1Member5', level1Member5Index], defaultReturn);
}

export function byLevel1Member6NthLevel2Member1(
  property,
  level1Member6Index = 0,
  defaultReturn,
) {
  return get(
    property,
    ['level1Member6', level1Member6Index, 'level2Member1'],
    defaultReturn,
  );
}

export function byLevel1Member6(property, defaultReturn) {
  return get(property, ['level1Member6'], defaultReturn);
}

export function byLevel1Member6Nth(
  property,
  level1Member6Index = 0,
  defaultReturn,
) {
  return get(property, ['level1Member6', level1Member6Index], defaultReturn);
}

export function byLevel1Member7DefinitionMember1(property, defaultReturn) {
  return get(property, ['level1Member7', 'definitionMember1'], defaultReturn);
}

export function byLevel1Member7(property, defaultReturn) {
  return get(property, ['level1Member7'], defaultReturn);
}

export function byLevel1Member8Level2Member1(property, defaultReturn) {
  return get(property, ['level1Member8', 'level2Member1'], defaultReturn);
}

export function byLevel1Member8Level2Member2(property, defaultReturn) {
  return get(property, ['level1Member8', 'level2Member2'], defaultReturn);
}

export function byLevel1Member8Level2Member3(property, defaultReturn) {
  return get(property, ['level1Member8', 'level2Member3'], defaultReturn);
}

export function byLevel1Member8Level2Member3Nth(
  property,
  level2Member3Index = 0,
  defaultReturn,
) {
  return get(
    property,
    ['level1Member8', 'level2Member3', level2Member3Index],
    defaultReturn,
  );
}

export function byLevel1Member8Level2Member4(property, defaultReturn) {
  return get(property, ['level1Member8', 'level2Member4'], defaultReturn);
}

export function byLevel1Member8Level2Member5(property, defaultReturn) {
  return get(property, ['level1Member8', 'level2Member5'], defaultReturn);
}

export function byLevel1Member8Level2Member6Level3Member1(
  property,
  defaultReturn,
) {
  return get(
    property,
    ['level1Member8', 'level2Member6', 'level3Member1'],
    defaultReturn,
  );
}

export function byLevel1Member8Level2Member6(property, defaultReturn) {
  return get(property, ['level1Member8', 'level2Member6'], defaultReturn);
}

export function byLevel1Member8(property, defaultReturn) {
  return get(property, ['level1Member8'], defaultReturn);
}

export function byLevel1Member9(property, defaultReturn) {
  return get(property, ['level1Member9'], defaultReturn);
}
"
`;
